File: soln-graph-problems/course-schedule-with-indegree.java
/*
https://leetcode.com/problems/course-schedule/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    static int[] indegree;
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int n=numCourses;
        ArrayList<Integer>[] adj_list=new ArrayList[n];
        for(int i=0;i<n;i++)
            adj_list[i]=new ArrayList<>();
        for(int edge[]:prerequisites)
            adj_list[edge[1]].add(edge[0]);
        
        indegree=new int[n];        
        for(ArrayList<Integer> nodes:adj_list){
            for(int v:nodes)
                indegree[v]++;
        }
        
        Queue<Integer> queue=new LinkedList<>();
        for(int i=0;i<n;i++){
            if(indegree[i]==0)
                queue.add(i);
        }
        
        int order=0;
        while(!queue.isEmpty()){
            int node=queue.remove();
            order++;
            for(int adj_node:adj_list[node]){
                indegree[adj_node]--;
                if(indegree[adj_node]==0)
                    queue.add(adj_node);
            }
        }
        
        return order==n;
    }
}

File: C++/soln-graph-problems/course-schedule-with-indegree.cpp
/**
 @author Farheen Bano
  
 Date 24-07-2021
 
 Reference-
 https://leetcode.com/problems/course-schedule/
*/

class Solution {
public:

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj_list(numCourses);
        vector<int> indegree(numCourses,0);   

        for(auto edge:prerequisites)
            adj_list[edge[1]].push_back(edge[0]);
        
        for(auto nodes:adj_list){
            for(int v:nodes)
                indegree[v]++;
        }
        
        queue<int> q;
        for(int i=0;i<numCourses;i++){
            if(indegree[i]==0)
                q.push(i);
        }
        
        int order=0;
        while(!q.empty()){
            int node=q.front();
            q.pop();
            order++;
            for(int adj_node:adj_list[node]){
                indegree[adj_node]--;
                if(indegree[adj_node]==0)
                    q.push(adj_node);
            }
        }
        
        return order==numCourses;
    }
};